crates/bevy_render/src/batching/gpu_preprocessing.rs

```diff
@@ -823,72 +831,162 @@ where
-    /// indirect draw parameters.
-    non_indexed_metadata: RawBufferVec<IndirectParametersMetadata>,
+    /// The `indexed` parameter specifies whether the meshes that these batches
+    /// correspond to are indexed or not.
+    fn batch_count(&self, indexed: bool) -> usize {
+        if indexed {
+            self.indexed.batch_count()
+        } else {
+            self.non_indexed.batch_count()
+        }
+    }
 
-    /// The GPU buffer that holds the number of indirect draw commands for each
-    /// phase of each view, for non-indexed meshes.
+    /// Returns the number of batch sets currently allocated.
     ///
-    /// The indirect parameters building shader writes to this buffer, and the
-    /// `multi_draw_indirect_count` command reads from it in order to know how
-    /// many indirect draw commands to process.
-    non_indexed_batch_sets: RawBufferVec<IndirectBatchSet>,
+    /// The `indexed` parameter specifies whether the meshes that these batch
+    /// sets correspond to are indexed or not.
+    pub fn batch_set_count(&self, indexed: bool) -> usize {
+        if indexed {
+            self.indexed.batch_sets.len()
+        } else {
+            self.non_indexed.batch_sets.len()
+        }
+    }
+
+    /// Adds a new batch set to `Self::indexed_batch_sets` or
+    /// `Self::non_indexed_batch_sets` as appropriate.
+    ///
+    /// `indexed` specifies whether the meshes that these batch sets correspond
+    /// to are indexed or not. `indirect_parameters_base` specifies the offset
+    /// within `Self::indexed_data` or `Self::non_indexed_data` of the first
+    /// batch in this batch set.
+    pub fn add_batch_set(&mut self, indexed: bool, indirect_parameters_base: u32) {
+        if indexed {
+            self.indexed.batch_sets.push(IndirectBatchSet {
+                indirect_parameters_base,
+                indirect_parameters_count: 0,
+            });
+        } else {
+            self.non_indexed.batch_sets.push(IndirectBatchSet {
+                indirect_parameters_base,
+                indirect_parameters_count: 0,
+            });
+        }
+    }
 
-    /// The GPU buffer that stores the indirect draw parameters for indexed
-    /// meshes.
+    /// Returns the index that a newly-added batch set will have.
+    ///
+    /// The `indexed` parameter specifies whether the meshes in such a batch set
+    /// are indexed or not.
+    pub fn get_next_batch_set_index(&self, indexed: bool) -> Option<NonMaxU32> {
+        NonMaxU32::new(self.batch_set_count(indexed) as u32)
+    }
+
+    /// Clears out the buffers in preparation for a new frame.
+    pub fn clear(&mut self) {
+        self.indexed.clear();
+        self.non_indexed.clear();
+    }
+}
+
+/// The buffers containing all the information that indirect draw commands use
+/// to draw the scene, for a single mesh class (indexed or non-indexed), for a
+/// single phase.
+pub struct MeshClassIndirectParametersBuffers<IP>
```

Comment:
While I'm not super keen on the word 'class', I don't have a better suggestion. I like how much gets ripped out of this.